src/par_iter/internal.rs

```diff
@@ -95,6 +111,22 @@ pub trait UnindexedConsumer<ITEM>: Consumer<ITEM> {
 pub trait UnindexedProducer: IntoIterator + Send + Sized {
     fn can_split(&self) -> bool;
     fn split(self) -> (Self, Self);
+
+    /// Iterate the producer, feeding each element to `folder`, and
+    /// stop when the folder is full (or all elements have been consumed).
+    ///
+    /// The provided implementation is sufficient for most iterables.
+    fn fold_with<F>(self, mut folder: F) -> F
+        where F: Folder<Self::Item>,
+    {
+        for item in self {
+            folder = folder.consume(item);
```

Comment:
So the key distinctions are:

- can split at an arbitrary point
    - needed for `zip`
- knows exactly how many elements it will produce
    - needed for precise consumers

Today we tie those together, but they could be separated. Something like:

- UnindexedProducer
    - can split itself at some point
- KnownLengthProducer: UnindexedProducer
    - can tell you the number of items it will produce
- Producer: KnownLengthProducer
    - can further be split at an arbitrary point

we might then have two bridge methods:

- one takes a `KnownLengthProducer` and a `Consumer`
    - this is the one that today requires a `Producer`
    - it would instead invoke `producer.split()`, get resulting length, and invoke `consumer.split_at()`, as you suggested
- one takes a `UnindexedProducer` and a `UnindexedConsumer`
    - same as today

We would probably have to tweak the `ParallelIterator` hierarchy a bit too, but the idea is that `zip()` would require that its sub-iterators can produce a full `Producer` (or, potentially, that the LHS can produce a `KnownLengthProducer` and the RHS can produce a full `Producer`). It's `split()` implementation can then call `left.split()` and right.split_at(left_split.len())`.

Make sense?